File: src/karkinolution/decisions/intent.py
from karkinolution.decisions.actions import Intent, IntentActs
from karkinolution.decisions.brain.instincts import (
    Instincts,
    try_call_reproductive_buffer,
)
from karkinolution.decisions.brain.planners import Planner
from karkinolution.decisions.perception import Perception
from karkinolution.decisions.presets import (
    AttackPreset,
    EatPreset,
    MovePreset,
    ReproducePreset,
)

from karkinolution.organism.creatures import Creature

from karkinolution.systems.reproductivebuffer import ReproductiveBuffer

from karkinolution.terrain.world import World

class IntentResolver:
    @staticmethod
    def to_nothing_intent(creature:Creature) -> None:
        creature.intent = Intent(IntentActs.NOTHING)
    @staticmethod
    def cancel_invalid_intents(creature:Creature) -> None:
        intent = creature.intent.intent
        intent_time = creature.intent.time
        if creature.hungry < creature.genome.metabolism.max_hungry and intent_time > 2 and intent == IntentActs.FIND_FOOD:
            IntentResolver.to_nothing_intent(creature)
        if intent_time > 5 and intent ==  IntentActs.FIND_MATCH:
            IntentResolver.to_nothing_intent(creature)


    @staticmethod
    def update_intent(creature:Creature, reproductive_buffer:ReproductiveBuffer) -> (
        MovePreset | EatPreset | AttackPreset | ReproducePreset | None):
        creature.intent.time += 1
        if creature.intent.intent == IntentActs.NOTHING:
            acts = Instincts.take(creature, reproductive_buffer)
            Instincts.apply_noise(acts)

            act = Instincts.chose(acts)
            creature.intent = act
            try_call_reproductive_buffer(act, creature, reproductive_buffer)




    @staticmethod
    def resolve_intent(creature:Creature, world:World, perception:Perception) -> MovePreset | AttackPreset | EatPreset | None:
        IntentResolver.cancel_invalid_intents(creature)
        # ALIAS
        reproductive_buffer = world.reproductive_buffer
        entities = world.entities
        # CODE
        IntentResolver.update_intent(creature, reproductive_buffer)
        return Planner.plan(perception, creature, entities, world.territory)

File: src/karkinolution/brain/intent.cpp
#include <karkinolution/brain/intent.hpp>
#include <karkinolution/organism/creatures.hpp>
#include <karkinolution/actions/actions.hpp>
#include <karkinolution/world/world.hpp>
#include <karkinolution/systems/reproductivebuffer.hpp>
#include <karkinolution/actions/presets.hpp>
#include <optional>
#include <karkinolution/brain/perception.hpp>
#include <karkinolution/brain/instincts/instincts.hpp>
#include <karkinolution/brain/planners.hpp>

namespace IntentResolver {

    void to_nothing_intent(Creature& creature) {
        creature.intent = Intent(IntentActs::NOTHING);
    }

    void cancel_invalid_intents(Creature& creature) {
        const IntentActs act = creature.intent.act;
        const int time = creature.intent.time;

        if (creature.hungry().value() < creature.genome.metabolism.max_hungry && time > 2 && act == IntentActs::FIND_FOOD) {
            IntentResolver::to_nothing_intent(creature);
        }

        if (time > 5 && act == IntentActs::FIND_MATCH) {
            IntentResolver::to_nothing_intent(creature);
        }
    }

    void update_intent(Creature& creature, ReproductiveBuffer& reproductive_buffer) {
        creature.intent.time += 1;
        if (creature.intent.act == IntentActs::NOTHING) {
            IntentActs act = Instincts::take(creature, reproductive_buffer);

            creature.intent = Intent(act);
            try_call_reproductive_buffer(creature.intent, creature, reproductive_buffer);
        }
    }

    std::optional<AllPresets> resolve_intent(
        Creature& creature, 
        World& world, 
        const Perception& perception, 
        const DangerIndex& danger_index
    ) {
        IntentResolver::cancel_invalid_intents(creature);
        IntentResolver::update_intent(creature, world.reproductive_buffer);

        std::optional<std::variant<FindFoodPresetsOutput, AttackPresetsOutput>> result =  Planner::plan(perception, danger_index, creature, world.entities);

        if (!result)
            return std::nullopt;

        return std::visit(
            [](const auto& variant) -> AllPresets {
                return std::visit(
                    [](const auto& preset) -> AllPresets {
                        return preset;
                    },
                    variant
                );
            },
            *result
        );
    }
} // namespace IntentResolver